kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -0,0 +1,125 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.util;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import static java.lang.String.format;
+
+import io.delta.kernel.types.*;
+
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+
+/**
+ * Utility methods for schema related operations such as validating the schema has no duplicate
+ * columns and the names contain only valid characters.
+ */
+public class SchemaUtils {
+
+    private SchemaUtils() {
+    }
+
+    /**
+     * Validate the schema. This method checks if the schema has no duplicate columns and the names
+     * contain only valid characters.
+     *
+     * @param schema                 the schema to validate
+     * @param isColumnMappingEnabled whether column mapping is enabled. When column mapping is
+     *                               enabled, the column names in the schema can contain special
+     *                               characters that are allowed as column names in the Parquet
+     *                               file
+     * @throws IllegalArgumentException if the schema is invalid
+     */
+    public static void validateSchema(StructType schema, boolean isColumnMappingEnabled) {
+        List<String> flattenColNames = flattenNestedFieldNames(schema);
+
+        // check there are no duplicate column names in the schema
+        Set<String> uniqueColNames = flattenColNames.stream()
+                .map(String::toLowerCase)
+                .collect(Collectors.toSet());
+
+        if (uniqueColNames.size() != flattenColNames.size()) {
+            String duplicateColumns = flattenColNames.stream()
+                    .map(String::toLowerCase)
+                    .filter(n -> !uniqueColNames.add(n))
+                    .collect(Collectors.joining(", "));
+            throw new IllegalArgumentException(
+                    "Schema contains duplicate columns: " + duplicateColumns);
+        }
+
+        // Check the column names are valid
+        if (!isColumnMappingEnabled) {
+            validParquetColumnNames(flattenColNames);
+        } else {
+            // when column mapping is enabled, just check the name contains no new line it.
+            flattenColNames.forEach(name -> checkArgument(
+                    !name.contains("\\n"),
+                    format("Attribute name '%s' contains invalid new line characters.", name)));
+        }
+    }
+
+    /**
+     * Returns all column names in this schema as a flat list. For example, a schema like: | - a | |
```

Comment:
I can't really easily read or understand `| - a | | - 1 | | - 2 | - b | - c | | - nest |   | - 3`

Can you make this multi line and in a tree format?

Reply:
This is because of Intellij auto-format. Added `<pre>` tag to avoid formatting this block.